quickwit/quickwit-storage/src/cache/byte_range_cache.rs

```diff
@@ -0,0 +1,359 @@
+            .filter(|(k, v)| k.tag == query.tag && range_end <= v.range_end)
+    }
+
+    pub fn record_item(&mut self, num_bytes: usize) {
+        self.num_items += 1;
+        self.num_bytes += num_bytes as u64;
+        self.cache_counters.in_cache_count.inc();
+        self.cache_counters.in_cache_num_bytes.add(num_bytes as i64);
+    }
+
+    pub fn drop_item(&mut self, num_bytes: usize) {
+        self.num_items -= 1;
+        self.num_bytes -= num_bytes as u64;
+        self.cache_counters.in_cache_count.dec();
+        self.cache_counters.in_cache_num_bytes.sub(num_bytes as i64);
+    }
+}
+
+impl<T: 'static + ToOwned + ?Sized> Drop for NeedMutByteRangeCache<T> {
+    fn drop(&mut self) {
+        self.cache_counters
+            .in_cache_count
+            .sub(self.num_items as i64);
+        self.cache_counters
+            .in_cache_num_bytes
+            .sub(self.num_bytes as i64);
+    }
+}
+
+/// Cache for ranges of bytes in files.
+/// Contrary to `MemorySizedCache`, it's able to answer subset of known ranges.
+///
+/// This cache assume immutable data: if you put a new slice and it overlap with
+/// cached data, the changes may or may not get recorded.
+// At the moment this is hardly a cache as it features no eviction policy.
+pub struct ByteRangeCache {
+    inner: Mutex<NeedMutByteRangeCache<Path>>,
+}
+
+impl ByteRangeCache {
+    /// Creates a slice cache that nevers removes any entry.
+    pub fn with_infinite_capacity(cache_counters: &'static CacheMetrics) -> Self {
+        ByteRangeCache {
+            inner: Mutex::new(NeedMutByteRangeCache::with_infinite_capacity(
+                cache_counters,
+            )),
+        }
+    }
+
+    /// If available, returns the cached view of the slice.
+    pub fn get_slice(&self, path: &Path, byte_range: Range<usize>) -> Option<OwnedBytes> {
+        self.inner.lock().unwrap().get_slice(path, byte_range)
+    }
+
+    /// Put the given amount of data in the cache.
+    pub fn put_slice(&self, path: PathBuf, byte_range: Range<usize>, bytes: OwnedBytes) {
+        self.inner
+            .lock()
+            .unwrap()
+            .put_slice(path, byte_range, bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+    use std::ops::Range;
+    use std::path::Path;
+
+    use proptest::prelude::*;
+
+    use super::ByteRangeCache;
+    use crate::metrics::CACHE_METRICS_FOR_TESTS;
+    use crate::OwnedBytes;
+
+    #[derive(Debug)]
+    enum Operation {
+        Insert {
+            range: Range<usize>,
+            tag: &'static str,
+        },
+        Get {
+            range: Range<usize>,
+            tag: &'static str,
+        },
+    }
+
+    fn tag_strategy() -> impl Strategy<Value = &'static str> {
+        prop_oneof![Just("path1"), Just("path2"),]
+    }
+
+    fn range_strategy() -> impl Strategy<Value = Range<usize>> {
+        (0usize..4095usize).prop_perturb(|start, mut rng| start..rng.gen_range(start..4096))
```

Comment:
This 4096 is way too high and prevents you from covering much ground.

```suggestion
        (0usize..11usize).prop_perturb(|start, mut rng| start..rng.gen_range(start..12usize))
```